fix loading in cast credit

mohamedebrahem13 opened the pull request:
- fix loading in cast credit

mohamedebrahem13 committed:
fix loading in cast credit

poula22 commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt:
```diff
@@ -50,13 +50,16 @@ class CastCreditViewModel @Inject constructor(
             onSuccess = ::updateCastCreditPosters,
             onError = ::loadCastCreditError
         ) {
+            val allSeriesGenres = getSeriesGenresByIds(emptyList())
+            val allMoviesGenres = getMoviesGenresByIds(emptyList())
```
This can cause race condition. it should be hold in stateFlow

mohamedebrahem13 committed:
add genders list to the sate

mohamedebrahem13 committed:
Merge branch 'develop' into fix/loading-in-cast-credit

mohamedebrahem13 committed:
fix conflict

skaik-mo commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt:
```diff
@@ -33,8 +33,22 @@ class CastCreditViewModel @Inject constructor(
-        state.value.castId?.let {
-            loadCastCredit(it)
+
+        safeExecute(
```
Separating the safeExecute logic into its own function for better readability. Also, it would be clearer if the onSuccess block is extracted into a separate function.

skaik-mo commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt:
```diff
@@ -33,8 +33,22 @@ class CastCreditViewModel @Inject constructor(
+            onError = ::loadCastCreditError
+        ) {
+            val series = getSeriesGenresByIds(emptyList())
+            val movies = getMoviesGenresByIds(emptyList())
```
Use `safeExecute` for each use case to ensure consistency and safer execution handling.

skaik-mo replied:
Also, using `ObserveSeriesGenresUseCase` and `ObserveMoviesGenresUseCase` instead of `GetSeriesGenresByIdsUseCase` and `GetMoviesGenresByIdsUseCase`

skaik-mo commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt:
```diff
@@ -54,19 +68,22 @@ class CastCreditViewModel @Inject constructor(
+                val genres = it.genreIDs.mapNotNull { id ->
+                    state.value.allSeriesGenres.find { g -> g.id == id }?.title
+                }
                 it.toUi().toPoster().copy(genres = genres.joinToString(", "))
```
`joinToString(", ")` is already handled inside toPoster, so it can be removed here to avoid duplication.

skaik-mo commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt:
```diff
@@ -54,19 +68,22 @@ class CastCreditViewModel @Inject constructor(
+                val genres = it.genresID.mapNotNull { id ->
+                    state.value.allMovieGenres.find { g -> g.id == id }?.title
+                }
                 it.toUi().toPoster().copy(genres = genres.joinToString(", "))
```
also here
> joinToString(", ") is already handled inside toPoster, so it can be removed here to avoid duplication.

skaik-mo reviewed (changes requested):
After fixing the bugs, the genres should appear as expected. Ensure they are displayed correctly.

https://github.com/user-attachments/assets/884b69dc-8986-4ab1-943f-f79433f0a0b1

mohamedebrahem13 committed:
use observes genres use case instead of get by ids , separate the safe execute

skaik-mo commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt:
```diff
@@ -54,19 +80,22 @@ class CastCreditViewModel @Inject constructor(
             onError = ::loadCastCreditError
```
Remove `safeExecute` here, it’s not needed in this context.

mohamedebrahem13 committed:
remove un used safe execute

skaik-mo committed:
Refactor: Improve CastCreditScreen data loading

- Removed `onRetryClick` functionality as it's handled by `BaseScreen`.
- Modified data loading to observe genres and then fetch cast credits.
- Updated `CastCreditScreenState` to reflect loading state more accurately.
- Ensured distinct posters and sorted them by date.

skaik-mo committed:
Merge branch 'fix/loading-in-cast-credit' of github.com:Rome-Squad/CineVerse-App into fix/loading-in-cast-credit

* 'fix/loading-in-cast-credit' of github.com:Rome-Squad/CineVerse-App:
  remove un used safe execute

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditViewModel.kt

HendSayed25 commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditInteractionListener.kt:
```diff
@@ -7,6 +7,4 @@ interface CastCreditInteractionListener {
 
     fun changeView(isGrid: Boolean)
-
-    fun onRetryClick()
```
Why remove this function???

mohamedebrahem13 replied:
@skaik-mo Ask

skaik-mo replied:
We don’t need `onRetryClick`. It’s meant for re-requesting `loadCastCredit` when a cast ID is not found, but since we always have a valid cast ID, this case will never occur.

HendSayed25 commented on presentation/details/src/main/java/com/giraffe/presentation/details/screens/castCredit/CastCreditScreen.kt:
```diff
@@ -50,7 +50,6 @@ private fun CastCreditContent(
     BaseScreen(
         isLoading = state.isLoading,
         isNoInternet = state.isNoInternet,
-        onRetryClick = interaction::onRetryClick,
```
This used  to handle the case when disconnect the internet and return again , why remove it ?

mohamedebrahem13 replied:
Ask @skaik-mo

skaik-mo replied:
Now I understand why we need `onRetryClick`. It makes sense to keep it for handling cases when there is no internet connection.

skaik-mo committed:
Fix: Handle retry click in CastCreditScreen

This commit introduces the ability to retry loading data in the CastCreditScreen when a network error occurs.

Specifically, it:
- Adds an `onRetryClick` function to `CastCreditInteractionListener` and its implementation in `CastCreditViewModel`.
- Passes the `onRetryClick` callback to the `BaseScreen` in `CastCreditScreen`.
- Updates `CastCreditViewModel` to reset `isNoInternet` and set `isLoading` to true when retrying.
- Fixes a bug in the `toMedia` mapper where `firstCreditAirDate` was used for `releaseYear` instead of `releaseDate`.
